### ISSUE TITLE
CUDA OOM during validation with full bag size (large slides/tile sets exceed GPU memory) in train/crossval

### ISSUE THREAD

**Issue by ferdinand-popp** [KatherLab/STAMP#78]
When running STAMP cross-validation with the CONCH model embeddings (of one TCGA H&E cancer entity) on a 32 GB GPU, we encounter a CUDA out-of-memory error during validation — usually at the prediction step. This happens most likely due to slides with a large number of tiles (e.g. 20000,512 dim), as all tiles are included (full bag size), and the Vision Transformer model requires a lot of memory, especially due to multi-head attention.

We’re aware that reducing bag size in validation and test data loader could prevent this, but simple tile sampling might introduce bias or miss relevant regions. Potential ideas could be:

A smarter way of selecting or filtering tiles at validation/test time.
Skipping slides with extreme tile counts.
Or even offering a lighter model alternative for inference.
Let us know if you have suggestions — happy to provide logs or test potential fixes.

**Comment by georg-wolflein** [comment]
There are two possible reasons I can think of:
- we could get OOM due to the increased dimension of the feedforward layers. STAMP v1 had 512 (iirc; @s1787956 is this true?) and STAMP v2 has 2048
  https://github.com/KatherLab/STAMP/blob/4db5a74cbe214e5ab90bd1c5d5a552be9acd037c/src/stamp/modeling/train.py#L292
- maybe we are we computing gradients at deployment? (We shouldn't need to...) Perhaps wrapping the following lines in a `with torch.no_grad()` might help:
  https://github.com/KatherLab/STAMP/blob/4db5a74cbe214e5ab90bd1c5d5a552be9acd037c/src/stamp/modeling/deploy.py#L171-L184

@s1787956 do you have any other ideas?

**Comment by EzicStar** [comment]
Hello @ferdinand-popp ,
we found that doing masking was causing a huge increase of memory usage during validation and deployment. I hope this fixes your issue, if not feel free to reopen the issue again. Other alternatives for optimizing the modeling are commented on #82.

**Comment by ferdinand-popp** [comment]
Dear @EzicStar, dear @FWao , dear @georg-wolflein,

thank you for the fix and the great support!

Best,
Ferdinand
